Marlin/src/module/planner.cpp

```diff
@@ -826,7 +827,7 @@ void Planner::calculate_trapezoid_for_block(block_t * const block, const_float_t
 
   // Store new block parameters
   block->accelerate_until = accelerate_steps;
-  block->decelerate_after = accelerate_steps + plateau_steps;
+  block->decelerate_after = decelerate_steps + plateau_steps;
```

Comment:
This doesn't look right to me. I suspect it might provide proper results in simple test cases with near-symmetrical accel/decel, but that won't always be the case.

If another segment follows this with a different speed than the incoming initial rate, you would end up decelerating either too early or too late.

It's possible I'm completely misunderstanding this code, so feel free to correct me if I'm wrong. If the new code is correct perhaps some variable name changes are in order to make it clear!

Reply:
I was just playing with this and discovered that in the `(plateau_steps < 0)` caee above `decelerate_steps` doesn't get updated. I just ended up with this from a 16-step move:

![image](https://user-images.githubusercontent.com/20053467/143731905-a6f027f5-bd62-44eb-8652-fbf48bac651d.png)

It looks like `decelerate_steps` is calculated _only_ to calculate plateau_steps, and isn't necessarily valid at this point in the code.

Reply:
How I came to this was knowledge is based on actual logged SERIAL_ECHO counts of all steps. 
You would actually decelerate to late in the current code and never to early. I saw no issue with the accel step count. 

From observing hundreds of real moves the logs always showed the actual stepper decel were shorted because the stepper uses the total step count to stop which happens multiple steps before it should based of the planner decel calc.

Reply:
I'm thinking of a trapezoid like this. Wouldn't your new form set `decelerate_after` to 6, which is still in the middle of the acceleration phase?

It wouldn't actually stop accelerating early (because it is in an if/else if), but it would start decelerating at the beginning of the plateau.

![image](https://user-images.githubusercontent.com/20053467/143732717-5ca96a3d-ffb3-40e6-a88e-a4f30aa9d16f.png)

Reply:
Hmm, it does not set a decel value when it cannot reach the final rate. I would think that is a defect, 16 steps would rarely reach the final rate and there is a test for that in the code. So that is why the accel was reused and it results in a short decel cycle every time the exit rate is lower than the cruise rate.

Reply:
Well not every time, just when the calc is odd valued.

Reply:
When it comes down to it, this is pretty hard code to get right without some really structured testing around it. If this planning/stepper logic could be unit tested I can pretty much guarantee it would be riddled with bugs.

I say this not to criticize the code, but from my own past experience writing unit tests for what was understood to be "correct" code in other projects.

Reply:
This resolves the issue where the nominal rates cannot be reached:
decelerate_steps = block->step_event_count - accelerate_steps;

Reply:
That solves the issue with `decelerate_steps` is invalid. I still cannot understand how `decelerate_steps + plateau_steps` is the right point to begin deceleration, but we can't seem to converge our thoughts on that particular concern.

Reply:
How is it invalid?

Reply:
I see the problem now, if the final rate is not symmetrical it will be incorrect.
We must base the decelerate_after value for the end point and that should resolve the odd even problem.

block->decelerate_after =  block->step_event_count - decelerate_steps;